modules/standard/OS.chpl

```diff
@@ -1298,53 +1328,135 @@ module OS {
   }
 
   /*
-     :class:`IoError` is the subclass of :class:`SystemError` that serves as the
-     base class for all errors regarding inputs and their formatting.
-     They are typically not directly generated by the OS, but they are
-     used and emitted by the IO module.
+     :class:`IoError` is the subclass of :class:`SystemError` corresponding to
+     EIO.
   */
   class IoError : SystemError {
-    proc init(err: errorCode, details: string = "") {
+    proc init(err: errorCode = EIO, details: string = "") {
       super.init(err, details);
     }
   }
 
   /*
-     :class:`EofError` is the subclass of :class:`IoError` corresponding to
+     :class:`EofError` is the the Chapel-specific error corresponding to
      encountering end-of-file.
   */
-  class EofError : IoError {
-    proc init(details: string = "", err: errorCode = EEOF:errorCode) {
-      super.init(err, details);
+  class EofError : Error {
+    var details: string;
+    var err_msg: string;
+
+    proc init(details: string = "", err_msg: string = "") {
+      this.details = details;
+      this.err_msg = err_msg;
+    }
+
+    override proc message() {
+      var msg: string;
+
+      if !err_msg.isEmpty() {
+        msg += err_msg;
+      } else {
+        // use default err_msg based on error code
+        var err:errorCode = EEOF;
+        var strerror_err_ignore: c_int = 0;
+        var errstr = sys_strerror_syserr_str(err, strerror_err_ignore);
+        var errorcode_msg: string;
+        try! {
+          errorcode_msg = createStringWithOwnedBuffer(errstr);
+        }
+        msg += errorcode_msg;
+      }
+
+      // add details if present
+      if !details.isEmpty() then
+        msg += " (" + details + ")";
+
+      return msg;
     }
   }
 
   /*
-     :class:`UnexpectedEofError` is the subclass of :class:`IoError`
+     :class:`UnexpectedEofError` is the Chapel-specific error
      corresponding to encountering end-of-file before the requested amount of
      input could be read.
   */
```

Comment:
I might also add: "This error can also occur on some writing operations when a :record:`~IO.fileWriter`'s range has been specified, and the write exceeds the valid range.